### PR TITLE
implement force removal of packages

### LINKED ISSUE #9: force remove Numba and llvmlite
Before installing a source add the ability to force remove and existing installs of the source and any important dependencies. This may happen when target dependencies include the source.

### PR DESCRIPTION
**Pull request by esc**
As detailed in the docstrings, sometimes we need to remove certain
packages before installing from a git clone. This implements the
required wiring by provding an 'abstract' property and a means of
running `conda remove --force`.

Fixes #9

### CODE REVIEW

File: texasbbq.py [hunk 1]
```diff
@@ -215,6 +215,25 @@ def conda_dependencies(self):
         """
         raise NotImplementedError
 
+    @property
+    def conda_force_remove_packages(self):
+        """Conda packages to remove before installation.
+
+        Sometimes, the dependencies of as source will include the source itself
```
Comment (stuartarchibald) [inline texasbbq.py:222]:
```suggestion
        Sometimes, the dependencies of a source will include the source itself
```

File: texasbbq.py [hunk 2]
```diff
@@ -215,6 +215,25 @@ def conda_dependencies(self):
...
+        (and potentially other critical dependencies). This will mean, that a
```
Comment (stuartarchibald) [inline texasbbq.py:223]:
```suggestion
        (and potentially other critical dependencies). This means that a
```

File: texasbbq.py [hunk 3]
```diff
@@ -215,6 +215,25 @@ def conda_dependencies(self):
...
+        released version of the source may be installed and this must be
+        removed before the source can be installed in the correct version. This
```
Comment (stuartarchibald) [inline texasbbq.py:225]:
```suggestion
        removed before the current version source can be installed. This
```

File: texasbbq.py [hunk 4]
```diff
@@ -249,6 +268,18 @@ def install(self, env):
             execute("conda run -n {} {}".format(env, self.install_command))
             os.chdir('../')
 
+    def conda_force_remove(self, env):
+        """Execute command to force remove any packages before installation.
+
+        Parameters
+        ----------
+        env: str
+            The conda environment to remove packages from.
```
Comment (stuartarchibald) [inline texasbbq.py:277]:
```suggestion
            The conda environment from which to remove packages.
```

File: texasbbq.py [hunk 5]
```diff
@@ -249,6 +268,18 @@ def install(self, env):
...
+
+        """
+        for p in self.conda_force_remove_packages:
+            execute("conda remove -n {} -y --force {} ".format(env, p))
```
Comment (stuartarchibald) [inline texasbbq.py:281]: does force remove handle multiple packages at once? I don't recall.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. Few minor things else looks good.

**Comment by esc** [comment]
So, the problem with the approach is the following: force removal will leave the environment in an inconsistent state. That means, any new install commands will not only install the package given on the command line, but also attempt to fix the inconsistent environment. So for example any attempt to install only `llvmlite`  after  force removing both `llvmlite` and `numba` will mean that  `numba` will be installed as a side-effect in order to fix the  environment. In the Numba case this means that a development  build from  anaconda.org may override a subsequent install from a git checkout. The bottom line is that the approach in this PR is unlikely to succeed. 

Perhaps an alternative approach could be to install the dependencies of the source and the source into the environment BEFORE installing any targets and their dependencies. However, in that case it might be wise to check before running the tests that the original version of source and source dependencies are still installed.

**Comment by esc** [comment]
#9 is no longer needed so this can be closed too. See #20 https://github.com/numba/numba-integration-testing/pull/50 instead which solve the problem as suggested in this thread.

**Pull request closed without merging**